examples/context-propagation/rxjava/src/main/java/example/armeria/contextpropagation/rxjava/MainService.java

```diff
@@ -0,0 +1,88 @@
+import com.linecorp.armeria.server.HttpService;
+import com.linecorp.armeria.server.ServiceRequestContext;
+
+import io.reactivex.Observable;
+import io.reactivex.Single;
+
+public class MainService implements HttpService {
+
+    private static final Splitter NUM_SPLITTER = Splitter.on(',').trimResults().omitEmptyStrings();
+
+    private final WebClient backendClient;
+
+    public MainService(WebClient backendClient) {
+        this.backendClient = requireNonNull(backendClient, "backendClient");
+    }
+
+    @Override
+    public HttpResponse serve(ServiceRequestContext ctx, HttpRequest req) {
+        // This logic mimics using a blocking method, which would usually be something like a MySQL
+        // database query using JDBC.
+        final CompletableFuture<List<Long>> fetchFromFakeDb = CompletableFuture.supplyAsync(
+                () -> {
+                    // The context is mounted in a thread-local, meaning it is available to all logic such
+                    // as tracing.
+                    checkState(ServiceRequestContext.current() == ctx);
+
+                    Uninterruptibles.sleepUninterruptibly(Duration.ofMillis(50));
+                    return ImmutableList.of(23L, -23L);
+                },
+                // Always run blocking logic on the blocking task executor. By using
+                // ServiceRequestContext.blockingTaskExecutor, you also ensure the context is mounted inside the
+                // logic (e.g., your DB call will be traced!).
+                ctx.blockingTaskExecutor());
+
+        final Single<HttpResponse> response = FutureConverter
+                .toSingle(req.aggregate())
+                .flatMapObservable(request -> {
+                    // The context is mounted in a thread-local, meaning it is available to all logic
+                    // such as tracing.
+                    checkState(ServiceRequestContext.current() == ctx);
+
+                    final List<Long> nums = new ArrayList<>();
+                    for (String token : Iterables.concat(
+                            NUM_SPLITTER.split(request.path().substring(1)),
+                            NUM_SPLITTER.split(request.contentUtf8()))) {
+                        nums.add(Long.parseLong(token));
+                    }
+
+                    return Observable.fromIterable(nums);
+                })
+                .mergeWith(FutureConverter.toSingle(fetchFromFakeDb)
+                                          .flatMapObservable(Observable::fromIterable))
+                .flatMapSingle(num -> {
+                    // The context is mounted in a thread-local, meaning it is available to all logic
+                    // such as tracing.
+                    checkState(ServiceRequestContext.current() == ctx);
+
+                    return FutureConverter.toSingle(backendClient.get("/square/" + num).aggregate());
+                })
+                .map(AggregatedHttpResponse::contentUtf8)
```

Comment:
One aspect of rxjava integration I noticed is methods like these are fairly ridiculously slow due to context mounting. An inevitable downside of having the integration be a single static method invocation in `main` I guess.

Reply:
By the way, I was able to confirm that the pattern with `subscribeOn` instead of using `RequestContextAssembly` does not have the performance issue. The key point I guess is `RequestContextSubscriber` mounts on every call to `onNext`

https://github.com/line/armeria/blob/master/rxjava/src/main/java/com/linecorp/armeria/common/rxjava/RequestContextSubscriber.java#L38

which ends up being called from every single callback in the flow.

Subscribing on the executor though mounts the context when the thread starts to process the flow, which ends up processing as many items as it can at the time. So if the buffer has 10 items, it will go through all 10 in a single invocation of the executor, meaning just a single threadlocal push/pop for 10 invocations of `onNext`.

Just an observation, I think the assembly is still useful since it is quite nice for a single method call to be able to guarantee callbacks have context, but we do have to be aware of the cost of the convenience.